### PR TITLE
Personal/dougflick/vars with performance

### PR DESCRIPTION
**Pull request by Flickdm**
## Description

Adds DEBUG Logs to print variable data, and adds functionality to measure variable performance.

- [ ] Breaking change?
  - Will this change break pre-existing builds or functionality without action being taken?
No this will not cause any breaking changes

## How This Was Tested

<_Please describe the test(s) that were run to verify the changes._>
Manual testing was performed to verify debug prints and performance measurements are accurate

## Integration Instructions

<_Describe how these changes should be integrated. Use N/A if nothing is required._>
N/A

### CODE REVIEW

File: MdeModulePkg/Library/DxeCorePerformanceLib/DxeCorePerformanceLib.c [hunk 1]
```diff
@@ -1214,6 +1214,28 @@ InsertFpdtRecord (
 
       break;
 
+    // MU_CHANGE
+    case PERF_VARIABLE_START_ID:
+    case PERF_VARIABLE_END_ID:
+      if ((String != NULL) && (AsciiStrLen (String) != 0)) {
```
Comment (apop5) [inline MdeModulePkg/Library/DxeCorePerformanceLib/DxeCorePerformanceLib.c:1220]: Replace AsciiStrlen with AsciiStrnLenS

File: MdeModulePkg/Library/PeiPerformanceLib/PeiPerformanceLib.c [hunk 2]
```diff
@@ -474,6 +477,28 @@ InsertFpdtRecord (
 
       break;
 
+    // MU_CHANGE
+    case PERF_VARIABLE_START_ID:
+    case PERF_VARIABLE_END_ID:
+      if ((String != NULL) && (AsciiStrLen (String) != 0)) {
```
Comment (apop5) [inline MdeModulePkg/Library/PeiPerformanceLib/PeiPerformanceLib.c:483]: Replace AsciiStrlen with AsciiStrnLenS

File: MdeModulePkg/Library/SmmCorePerformanceLib/SmmCorePerformanceLib.c [hunk 3]
```diff
@@ -716,6 +719,28 @@ InsertFpdtRecord (
 
       break;
 
+    // MU_CHANGE
+    case PERF_VARIABLE_START_ID:
+    case PERF_VARIABLE_END_ID:
+      if ((String != NULL) && (AsciiStrLen (String) != 0)) {
```
Comment (apop5) [inline MdeModulePkg/Library/SmmCorePerformanceLib/SmmCorePerformanceLib.c:725]: Replace AsciiStrlen with AsciiStrnLenS

**Review by apop5 — APPROVED** [review]

File: MdeModulePkg/Universal/Variable/RuntimeDxe/VariableStandaloneMm.inf [hunk 4]
```diff
@@ -82,6 +82,7 @@
   VariablePolicyLib
   VariablePolicyHelperLib
   MemoryTypeInfoSecVarCheckLib                 # MU_CHANGE TCBZ1086 - Mitigate potential system brick due to UEFI MemoryTypeInformation var changes
+  PerformanceLib
```
Comment (Javagedes) [inline MdeModulePkg/Universal/Variable/RuntimeDxe/VariableStandaloneMm.inf:85]: Missed a MU_CHANGE

File: MdeModulePkg/Universal/Variable/RuntimeDxe/VariableRuntimeDxe.inf [hunk 5]
```diff
@@ -78,6 +78,7 @@
   VariablePolicyHelperLib
   SafeIntLib
   MemoryTypeInfoSecVarCheckLib  # MU_CHANGE TCBZ1086 - Mitigate potential system brick due to UEFI MemoryTypeInformation var changes
+  PerformanceLib
```
Comment (Javagedes) [inline MdeModulePkg/Universal/Variable/RuntimeDxe/VariableRuntimeDxe.inf:81]: MU CHANGE

File: MdeModulePkg/Universal/Variable/RuntimeDxe/VariableSmm.inf [hunk 6]
```diff
@@ -86,6 +86,7 @@
   SafeIntLib
   AdvLoggerAccessLib                            ## MU_CHANGE
   MemoryTypeInfoSecVarCheckLib  # MU_CHANGE TCBZ1086 - Mitigate potential system brick due to UEFI MemoryTypeInformation var changes
+  PerformanceLib
```
Comment (Javagedes) [inline MdeModulePkg/Universal/Variable/RuntimeDxe/VariableSmm.inf:89]: MU_CHANGE

File: MdePkg/Include/Library/BaseLib.h [hunk 7]
```diff
@@ -1526,6 +1526,56 @@ UnicodeStrToAsciiStrS (
   IN      UINTN         DestMax
   );
 
+/**
```
Comment (Javagedes) [inline MdePkg/Include/Library/BaseLib.h:1529]: MU_CHANGE begin

File: MdePkg/Include/Library/BaseLib.h [hunk 8]
```diff
@@ -1526,6 +1526,56 @@ UnicodeStrToAsciiStrS (
...
+**/
+RETURN_STATUS
+EFIAPI
+ReplaceUnicodeStrToAsciiStrS (
+  IN      CONST CHAR16  *Source,
+  OUT     CHAR8         *Destination,
+  IN      UINTN         DestMax,
+  IN      CONST CHAR8   ReplacementChar
+);
+
```
Comment (Javagedes) [inline MdePkg/Include/Library/BaseLib.h:1578]: MU_CHANGE end

File: MdePkg/Include/Library/PerformanceLib.h [hunk 9]
```diff
@@ -40,7 +40,9 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #define PERF_INMODULE_END_ID       0x41
 #define PERF_CROSSMODULE_START_ID  0x50
 #define PERF_CROSSMODULE_END_ID    0x51
-
+// MU_CHANGE
+#define PERF_VARIABLE_START_ID     0x60
```
Comment (Javagedes) [inline MdePkg/Include/Library/PerformanceLib.h:44]: I would either do a MU_Change at the end of each line:

```
#define PERF_VARIABLE_START ID      0x60 # MU_CHANGE
```

Or do a MU_CHANGE begin, MU_CHANGE end

File: MdePkg/Include/Library/PerformanceLib.h [hunk 10]
```diff
@@ -348,6 +350,31 @@ LogPerformanceMeasurement (
   IN UINT32       Identifier
   );
 
+  /**
```
Comment (Javagedes) [inline MdePkg/Include/Library/PerformanceLib.h:353]: MU_CHANGE begin

File: MdePkg/Include/Library/PerformanceLib.h [hunk 11]
```diff
@@ -348,6 +350,31 @@ LogPerformanceMeasurement (
...
+RETURN_STATUS
+EFIAPI
+LogPerformanceMeasurementUnicode (
+  IN CONST VOID   *CallerIdentifier,
+  IN CONST VOID   *Guid     OPTIONAL,
+  IN CONST CHAR16 *String   OPTIONAL,
+  IN UINT64       Address  OPTIONAL,
+  IN UINT32       Identifier
+  );
+
```
Comment (Javagedes) [inline MdePkg/Include/Library/PerformanceLib.h:377]: MU_CHANGE end

File: MdePkg/Include/Library/PerformanceLib.h [hunk 12]
```diff
@@ -761,4 +788,36 @@ LogPerformanceMeasurement (
   Expression                   \
   PERF_CODE_END ()
 
+// MU_CHANGE
```
Comment (Javagedes) [inline MdePkg/Include/Library/PerformanceLib.h:791]: Begin

File: MdePkg/Include/Library/PerformanceLib.h [hunk 13]
```diff
@@ -761,4 +788,36 @@ LogPerformanceMeasurement (
...
+  then LogPerformanceMeasurement() is called.
+
+**/
+#define PERF_VARIABLE_END(Guid, Variable, Function)                                                                  \
+  do {                                                                                                               \
+    if (PerformanceMeasurementEnabled ()) {                                                                          \
+      LogPerformanceMeasurementUnicode (&gEfiCallerIdGuid, Guid, Variable, (UINT64)Function, PERF_VARIABLE_END_ID);  \
+    }                                                                                                                \
+  } while (FALSE)
+
```
Comment (Javagedes) [inline MdePkg/Include/Library/PerformanceLib.h:822]: MU_CHANGE end

File: MdePkg/Library/BaseLib/SafeString.c [hunk 14]
```diff
@@ -2690,6 +2690,119 @@ UnicodeStrToAsciiStrS (
   return RETURN_SUCCESS;
 }
 
+/**
```
Comment (Javagedes) [inline MdePkg/Library/BaseLib/SafeString.c:2693]: MU_CHANGE begin

File: MdePkg/Library/BaseLib/SafeString.c [hunk 15]
```diff
@@ -2690,6 +2690,119 @@ UnicodeStrToAsciiStrS (
...
+
+    *(Destination++);
+    *(Source++);
+  }
+
+  *Destination = '\0';
+
+  return RETURN_SUCCESS;
+}
+
```
Comment (Javagedes) [inline MdePkg/Library/BaseLib/SafeString.c:2805]: MU_CHANGE end

File: MdePkg/Library/BasePerformanceLibNull/PerformanceLib.c [hunk 16]
```diff
@@ -325,6 +325,34 @@ LogPerformanceMeasurement (
   return RETURN_SUCCESS;
 }
 
+/**
```
Comment (Javagedes) [inline MdePkg/Library/BasePerformanceLibNull/PerformanceLib.c:328]: MU_CHANGE begin

File: MdePkg/Library/BasePerformanceLibNull/PerformanceLib.c [hunk 17]
```diff
@@ -325,6 +325,34 @@ LogPerformanceMeasurement (
...
+  IN CONST VOID   *CallerIdentifier OPTIONAL,
+  IN CONST VOID   *Guid     OPTIONAL,
+  IN CONST CHAR16  *String   OPTIONAL,
+  IN UINT64       Address   OPTIONAL,
+  IN UINT32       Identifier
+  )
+{
+  return RETURN_SUCCESS;
+}
+
```
Comment (Javagedes) [inline MdePkg/Library/BasePerformanceLibNull/PerformanceLib.c:355]: MU_CHANGE end

**Review by Javagedes — APPROVED** [review]
Need to add appropriate MU_CHANGE tags throughout. I left comments on some,. but not all :)

**Comment by makubacki** [comment]
Please update the PR title

**Comment by makubacki** [comment]
@Flickdm if you can leave this open for one more day, I'd like to take a quick look through. I don't have time at this exact moment though.

**Comment by makubacki** [comment]
Can you please update the PR title, test, and integration instructions?

**Review by makubacki — COMMENTED** [review]
Have you thought about adding instrumentation to GetNextVariableName()? That is one of the more time consuming API calls across the total number of invocations.

File: MdeModulePkg/Universal/Variable/Pei/Variable.c [hunk 18]
```diff
@@ -1075,6 +1093,20 @@ PeiGetVariable (
...
   *DataSize = VarDataSize;
 
+// MU_CHANGE
+EXIT:
+
+  // MU_CHANGE
+  PERF_VARIABLE_END(VariableName, VariableName, PeiGetVariable);
+
+  // MU_CHANGE
+  DEBUG ((DEBUG_VARIABLE, "Exit: FunctionName(%a) VariableName(%s) VendorGuid(%g) Attributes(0x%x) Status(0x%x)\n",
```
Comment (makubacki) [inline MdeModulePkg/Universal/Variable/Pei/Variable.c:1103]: For status, it's best to use `%r`.

File: MdePkg/Library/BaseLib/SafeString.c [hunk 19]
```diff
@@ -2690,6 +2690,119 @@ UnicodeStrToAsciiStrS (
...
+                                   If PcdMaximumUnicodeStringLength is not zero,
+                                    and DestMax is greater than
+                                    PcdMaximumUnicodeStringLength. 
+                                   If DestMax is 0.
+  @retval RETURN_ACCESS_DENIED     If Source and Destination overlap.
+
+**/
+RETURN_STATUS
+EFIAPI
+ ReplaceUnicodeStrToAsciiStrS (
```
Comment (makubacki) [inline MdePkg/Library/BaseLib/SafeString.c:2736]: There's some minor formatting tweaks needed. Like an extra starting space here. Most of these should be found by uncrustify.

File: MdePkg/Include/Library/BaseLib.h [hunk 20]
```diff
@@ -1526,6 +1526,56 @@ UnicodeStrToAsciiStrS (
...
+                                   If PcdMaximumUnicodeStringLength is not zero,
+                                    and DestMax is greater than
+                                    PcdMaximumUnicodeStringLength.
+                                   If DestMax is 0.
+  @retval RETURN_ACCESS_DENIED     If Source and Destination overlap.
+
+**/
+RETURN_STATUS
+EFIAPI
+ReplaceUnicodeStrToAsciiStrS (
```
Comment (makubacki) [inline MdePkg/Include/Library/BaseLib.h:1572]: These changes to BaseLib should be added in a separate PR.

File: MdePkg/Include/Library/PerformanceLib.h [hunk 21]
```diff
@@ -40,7 +40,9 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #define PERF_INMODULE_END_ID       0x41
 #define PERF_CROSSMODULE_START_ID  0x50
 #define PERF_CROSSMODULE_END_ID    0x51
-
+// MU_CHANGE
```
Comment (makubacki) [inline MdePkg/Include/Library/PerformanceLib.h:43]: For the UEFI variable changes, I suggest two commits in one PR.

- Commit 1: Changes to PerformanceLib only - describe the changes in detail in commit message
- Commit 2: Changes to Universal/Variable - describe those changes in detail in commit message

In PR description, describe both changes.

File: MdeModulePkg/Universal/Variable/Pei/Variable.c [hunk 22]
```diff
@@ -1042,14 +1043,29 @@ PeiGetVariable (
...
+    __FUNCTION__,
+    PeiGetVariable,
+    VariableName,
+    VariableGuid,
+    Attributes,
+    (UINT64)*DataSize,
+    Data));
+
+  // MU_CHANGE
+  PERF_VARIABLE_BEGIN(VariableGuid, VariableName, PeiGetVariable);
```
Comment (makubacki) [inline MdeModulePkg/Universal/Variable/Pei/Variable.c:1057]: PEI normally has a fixed number of FPDT entries allowed and is relatively space constrained in pre-memory.

If you haven't already, please try this on a real platform and check the impact on FPDT record count in PEI. First boot can sometimes have more reads than subsequent boots so it'd be good to look at both.

**Comment by makubacki** [comment]
Overall, I think this is a good starting set of perf points for the variable drivers to get started.

Do you have a perf log and/or serial log from a public platform you can post to show output from the changes in the PR?

File: MdeModulePkg/Library/SmmPerformanceLib/SmmPerformanceLib.c [hunk 23]
```diff
@@ -19,6 +19,11 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/PcdLib.h>
 #include <Library/BaseMemoryLib.h>
 
+//
+// Data for FPDT performance records.
```
Comment (spbrogan) [inline MdeModulePkg/Library/SmmPerformanceLib/SmmPerformanceLib.c:23]: test comment on commit 9a3a
